Catalog3 3

slenbder opened the pull request:
## Проведена работа по задачам Модуля 3: Интеграция, доработка UI и тестирование

### Задачи:

- **Реализована функциональность кнопок и взаимодействие с пользователем**
    
    - _Описание:_
        - Реализовано добавление и удаление NFT в избранное при нажатии на кнопку "сердечко" в `NFTCollectionViewCell`.
        - Реализовано добавление и удаление NFT в корзину при нажатии на кнопку корзины.
        - Обновлено состояние кнопок в зависимости от текущего состояния (в избранном/в корзине).
        - Обновлена `CollectionViewModel` для управления состоянием избранного и корзины.
    - _Оценка:_ **5 часов**; _Факт:_ 6 часов.
    
- **Интеграция с бэкендом**
    
    - _Описание:_
        - Реализованы сетевые запросы для добавления и удаления NFT из избранного и корзины.
        - Обработаны ответы сервера и обновлены данные в приложении.
    - _Оценка:_ **2 часа**; _Факт:_ 5 часов.
    
- **Тестирование и отладка**
    
    - _Описание:_
        - Проведено полное тестирование приложения на разных устройствах и версиях iOS.
        - Исправлены найденные баги и оптимизирована производительность.
        - _Оценка:_ **1 час**; _Факт:_ 1 час.
    
- **Подготовка к ревью и написание документации**
    
    - _Описание:_
        - Обновлены комментарии в коде и обеспечена его читаемость.
    - _Оценка:_ **1 час**; _Факт:_ 1 час.

slenbder committed:
Реализовал обработку нажатий на кнопки корзины и лайка в ячейке. Подправил метод `configure`

slenbder committed:
Реализовал функционал лайков

slenbder committed:
Реализовал фнукционал взаимодействия с корзиной

slenbder committed:
Отполировал код пастой гои, добавил вкусные комментарии, расставил все по правильным местам и полочкам, привел все в порядок

slenbder committed:
Отредактировал декомпозицию

Kirchberg commented on FakeNFT/Scenes /Catalog/Catalog Screen/ViewModel/CatalogViewModel.swift:
```diff
@@ -78,4 +92,55 @@ final class CatalogViewModel: CatalogViewModelProtocol {
+    func getProfile(completion: @escaping () -> Void) {
+        let dispatchGroup = DispatchGroup()
+        
+        dispatchGroup.enter()
```
Использование `DispatchGroup` в данном случае избыточна, у нас только одна асинхронная таска

Kirchberg commented on FakeNFT/Scenes /Catalog/Collection Screen /View/CatalogDetailsScreenViewController.swift:
```diff
@@ -253,48 +252,65 @@ final class CatalogDetailsScreenViewController: UIViewController {
+
+// MARK: - NFTCollectionViewCellDelegate
+extension CatalogDetailsScreenViewController: NFTCollectionViewCellDelegate {
+    func tapLikeButton(with id: String) {
+        ProgressHUD.show()
+        view.isUserInteractionEnabled = false
+        viewModel.toggleLike(for: id) {
+            ProgressHUD.dismiss()
+            self.view.isUserInteractionEnabled = true
+        }
+    }
+    
+    func tapCartButton(with id: String) {
+        ProgressHUD.show()
+        view.isUserInteractionEnabled = false
+        viewModel.toggleCart(for: id) {
+            ProgressHUD.dismiss()
+            self.view.isUserInteractionEnabled = true
+        }
+    }
```
⚠️ При проверке выяснилось, что эти функции не срабатывают, поэтому при нажатии на сердечко и корзину не обновляются на бэке данные и соответственно не отображаются корректно в приложении

slenbder replied:
Исправил!

Kirchberg commented on FakeNFT/Foundation/MemoryStorage/ProfileStorage.swift:
```diff
@@ -0,0 +1,37 @@
+    
+    // MARK: - ProfileStorage Methods
+    func saveProfile(profile: Profile) {
+        syncQueue.async {
```
⚠️ Асинхронная очередь лишняя в этом классе, достаточно просто «сохранять профиль» без нее. Тут также могут быть гонки, из-за того, что мы в разных контекстах пытаемся получить и записать значение (см. проблему читателей и писателей: https://ru.wikipedia.org/wiki/Задача_о_читателях-писателях)

Чтобы ее исправить достаточно использовать NSLock (https://developer.apple.com/documentation/foundation/nslock)

Он не позволяет читателям и писателям одновременно получать доступ к ресурсу, что помогает сохранить правильное состояние в данном случае профиля. Необходимо это поправить, приведу пример:
```swift
private let mutex = NSLock()

func doSomething() {
    mutex.lock()

    // Выполняем работу с общим ресурсом
    print("Hello World")

    mutex.unlock()
}
```

slenbder replied:
Исправил!

Kirchberg commented on FakeNFT/Scenes /Catalog/Collection Screen /ViewModel/CollectionViewModel.swift:
```diff
@@ -83,4 +99,61 @@ final class CollectionViewModel: CollectionViewModelProtocol {
+            print("Добавили в массив лайков: \(nftId)")
+        }
+        
+        profileService.sendExamplePutRequest(likes: favoriteNFT, avatar: profile.avatar, name: profile.name) { [weak self] result in
```
Предлагаю уточнить название функции и внести изменения в ее нотацию. Сейчас слово "Example" может создавать впечатление, что это тестовая функция, а не часть основного функционала. Давайте подберем более описательное и ясное название, чтобы пользователям сервиса было проще понять его назначение.

slenbder replied:
Исправил!

slenbder committed:
Исправил критические ошибки указанные в Пулл-Реквесте:
- Приложение при проставлении лайка или корзины теперь обрабатывает сетевые запросы, и вызываются соответствующие методы делегата NFTCollectionViewCellDelegate: tapLikeButton и tapCartButton.
- В классе ProfileStorage исправил сохранение и получение профиля для устранения гонок в многопоточности.

slenbder committed:
Исправил дополнительные ошибки указанные в Пулл-Реквесте:

- Убрал `DispatchGroup` в `func getProfile`.
- Исправил название функции и внес изменения в ее нотацию.

Kirchberg commented on FakeNFT/Foundation/MemoryStorage/ProfileStorage.swift:
🔥

Kirchberg reviewed (approved):
Работа допущена до мерджа с основной веткой. Мердж должен происходить вместе со всей группой.

Возвращайся в следующей работе с вмердженными эпиками коллег. 
Эту часть работы нужно оформить в пул реквесте (develop->master), вся команда присылает на проверку одну и ту же ссылку. 

Нужно будет приложить скринкаст с демонстрацией своего эпика на симуляторе: 
рекомендую ссылку складывать не только в ПР, но и в Readme под описанием твоего эпика.
